### PR TITLE
refactor: Separate storage into its own package

### LINKED ISSUE #150: Separate Caddy module from certmagic Storage implementation
CertMagic does not have a dependency on Caddy so it should be possible to use `certmagic-gcs` without Caddy. Unfortunately because of [`init()`](https://github.com/grafana/certmagic-gcs/blob/main/module.go#L30-L32) it is not possible to use this package outside of Caddy. Separating out the CertMagic storage implementation from the Caddy module would be very helpful for non-Caddy use-cases.

Thank you :)

### PR DESCRIPTION
**Pull request by shrink**
Closes #150

I do not believe this is a breaking change because the Caddy module will continue to execute via the `init` entrypoint which now relies on the `storage` package.

I tested this branch in my non-Caddy CertMagic project using `replace` and it worked as a drop-in replacement:

```
replace github.com/grafana/certmagic-gcs => github.com/shrink/certmagic-gcs v0.0.0-20230730155505-0a3798824df6
```

After merge, the package can be added to the CertMagic wik: https://github.com/caddyserver/certmagic/wiki/Storage-Implementations

### CODE REVIEW

File: module.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/google/tink/go/aead"
 	"github.com/google/tink/go/insecurecleartextkeyset"
 	"github.com/google/tink/go/keyset"
+	certmagicgcs "github.com/grafana/certmagic-gcs/storage"
```
Comment (hairyhenderson) [inline module.go:14]: better to leave the package name as `storage`

```suggestion
	"github.com/grafana/certmagic-gcs/storage"
```

(and modify below)

**Review by hairyhenderson — COMMENTED** [review]
Thanks for this, @shrink - it looks useful!

Just one minor package name change, then this should be good to merge.

File: module.go (hunk 1 above)
Comment (shrink) [reply module.go]: @hairyhenderson good catch -- unsure why I added the alias 🤔 fixed!

**Review by hairyhenderson — APPROVED** [review]

**Comment by shrink** [comment]
@hairyhenderson re: the lint failure, I guess we have 3 options:

1. Rename the `storage` package to something else
2. Rename `StorageConfig` to `Config`
3. Add a linting exclusion for stuttering on `StorageConfig`

I guess, on reflection, since `Storage` is now its own package, `StorageConfig` is redundant and since `StorageConfig` isn't exposed for Caddy consumers renaming `StorageConfig` to `Config` would be backwards compatible.

That said, I don't have a particular preference -- happy to go with whatever approach is standard for this org.

**Comment by hairyhenderson** [comment]
> * Rename `StorageConfig` to `Config`

IMO that's the best approach - go ahead with that!

**Comment by shrink** [comment]
> IMO that's the best approach - go ahead with that!

done! The `golangci` tool passed locally so it should be good to re-run the Workflows now 🤞

**Review by hairyhenderson — APPROVED** [review]
Thanks @shrink!

**Pull request merged**